### PR TITLE
fix(appset): add debug logs around deleting ownerReferences and add warning docs about policy behavior

### PR DESCRIPTION
**Pull request by mikutas**
#12172

Previous PRs didn't fix policy behavior completely.
This PR is adding warning to doc about policy and some logs for debug.

Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [x] The title of the PR conforms to the [Toolchain Guide](https://argo-cd.readthedocs.io/en/latest/developer-guide/toolchain-guide/#title-of-the-pr)
* [ ] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [x] Does this PR require documentation updates?
* [x] I've updated documentation as required by this PR.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/blob/master/community/CONTRIBUTING.md#legal)
* [ ] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [x] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).
* [ ] My new feature complies with the [feature status](https://github.com/argoproj/argoproj/blob/master/community/feature-status.md) guidelines.
* [ ] I have added a brief description of why this PR is necessary and/or what this PR solves.
* [ ] Optional. My organization is added to USERS.md.
* [ ] Optional. For bug fixes, I've indicated what older releases this fix should be cherry-picked into (this may or may not happen depending on risk/complexity).

### CODE REVIEW

File: applicationset/controllers/applicationset_controller.go [hunk 1]
```diff
@@ -108,15 +108,19 @@ func (r *ApplicationSetReconciler) Reconcile(ctx context.Context, req ctrl.Reque
 
 	// Do not attempt to further reconcile the ApplicationSet if it is being deleted.
 	if applicationSetInfo.ObjectMeta.DeletionTimestamp != nil {
+		appsetName := applicationSetInfo.ObjectMeta.Name
+		logCtx.Debugf("DeletionTimestamp is set to %s", appsetName)
 		deleteAllowed := utils.DefaultPolicy(applicationSetInfo.Spec.SyncPolicy, r.Policy, r.EnablePolicyOverride).AllowDelete()
 		if !deleteAllowed {
+			logCtx.Debugf("ApplicationSet policy allows to delete")
```
Comment (pasha-codefresh) [inline applicationset/controllers/applicationset_controller.go:115]: i think here should "not allow to delete"

**Review by pasha-codefresh — CHANGES_REQUESTED** [review]

File: applicationset/controllers/applicationset_controller.go [hunk 2]
```diff
@@ -108,15 +108,19 @@ func (r *ApplicationSetReconciler) Reconcile(ctx context.Context, req ctrl.Reque
...
 		if !deleteAllowed {
+			logCtx.Debugf("ApplicationSet policy does not allows to delete")
 			if err := r.removeOwnerReferencesOnDeleteAppSet(ctx, applicationSetInfo); err != nil {
 				return ctrl.Result{}, err
 			}
-			controllerutil.RemoveFinalizer(&applicationSetInfo, argov1alpha1.ResourcesFinalizerName)
-			if err := r.Update(ctx, &applicationSetInfo); err != nil {
-				return ctrl.Result{}, err
-			}
+			logCtx.Debugf("ownerReferences is deleted from %s", appsetName)
```
Comment (ishitasequeira) [inline applicationset/controllers/applicationset_controller.go:119]: nit: We are not removing ownerReferences from the appSet but instead from the applications. Maybe reword the log message to something like `removing ownerReferences of appSetName from generated applications`.

File: applicationset/controllers/applicationset_controller.go [hunk 3]
```diff
@@ -108,15 +108,19 @@ func (r *ApplicationSetReconciler) Reconcile(ctx context.Context, req ctrl.Reque
 
 	// Do not attempt to further reconcile the ApplicationSet if it is being deleted.
 	if applicationSetInfo.ObjectMeta.DeletionTimestamp != nil {
+		appsetName := applicationSetInfo.ObjectMeta.Name
+		logCtx.Debugf("DeletionTimestamp is set to %s", appsetName)
 		deleteAllowed := utils.DefaultPolicy(applicationSetInfo.Spec.SyncPolicy, r.Policy, r.EnablePolicyOverride).AllowDelete()
 		if !deleteAllowed {
+			logCtx.Debugf("ApplicationSet policy does not allows to delete")
```
Comment (ishitasequeira) [inline applicationset/controllers/applicationset_controller.go:115]:
```suggestion
			logCtx.Debugf("ApplicationSet policy does not allow to delete")
```

**Review by ishitasequeira — APPROVED** [review]

**Comment by mikutas** [comment]
`Run unit tests for Go packages` failed with network issue
https://github.com/argoproj/argo-cd/actions/runs/8476154916/job/23225349437?pr=17357
image

**Comment by mikutas** [comment]
`Check changes to generated code` will be fixed in #17662

File: applicationset/controllers/applicationset_controller.go [hunk 4]
```diff
@@ -108,15 +108,19 @@ func (r *ApplicationSetReconciler) Reconcile(ctx context.Context, req ctrl.Reque
 
 	// Do not attempt to further reconcile the ApplicationSet if it is being deleted.
 	if applicationSetInfo.ObjectMeta.DeletionTimestamp != nil {
+		appsetName := applicationSetInfo.ObjectMeta.Name
+		logCtx.Debugf("DeletionTimestamp is set to %s", appsetName)
```
Comment (ishitasequeira) [inline applicationset/controllers/applicationset_controller.go:112]: nit:
```suggestion
		logCtx.Debugf("DeletionTimestamp is set on %s", appsetName)
```

File: docs/operator-manual/applicationset/Controlling-Resource-Modification.md [hunk 5]
```diff
@@ -77,6 +77,22 @@ spec:
     applicationsSync: create-update
 ```
 
+### How to prevent Application controller from deleting Applications when deleting ApplicationSet
```
Comment (ishitasequeira) [inline docs/operator-manual/applicationset/Controlling-Resource-Modification.md:80]: The policy descriptions above at lines 42 - 63 state that the policy setting prevents modifications/deletion of applications, and we are stating here that the policy setting is not effective to prevent the deletion. I think this is causing some inconsistency in the documentation. 

Can we add a line in above sections for pointing to this section or incorporate this section in each of the above sections? WDYT?

**Review by ishitasequeira — COMMENTED** [review]
@mikutas , I had approved the PR earlier but I just realized some inconsistency in the documentation.

File: docs/operator-manual/applicationset/Controlling-Resource-Modification.md (hunk 5 above)
Comment (mikutas) [reply docs/operator-manual/applicationset/Controlling-Resource-Modification.md:86]: Added a line for pointing to this section.

**Review by gdsoumya — APPROVED** [review]
LGTM

**Comment by mikutas** [comment]
@pasha-codefresh please confirm that requested change is adressed.

**Comment by mikutas** [comment]
closing PR because its' unable to progress
